packages/mdc-list/foundation.js

```diff
@@ -0,0 +1,196 @@
+   */
+  handleFocusOut(evt) {
+    const listItem = this.getListItem_(evt.target);
+    if (!listItem) return;
+
+    this.adapter_.setTabIndexForListItemChildren(this.adapter_.getListItemIndex(listItem), -1);
+  }
+
+  /**
+   * Key handler for the list.
+   * @param {Event} evt
+   */
+  handleKeydown(evt) {
+    const arrowLeft = evt.key === 'ArrowLeft' || evt.keyCode === 37;
+    const arrowUp = evt.key === 'ArrowUp' || evt.keyCode === 38;
+    const arrowRight = evt.key === 'ArrowRight' || evt.keyCode === 39;
+    const arrowDown = evt.key === 'ArrowDown' || evt.keyCode === 40;
+    const isHome = evt.key === 'Home' || evt.keyCode === 36;
+    const isEnd = evt.key === 'End' || evt.keyCode === 35;
+    let currentIndex = this.adapter_.getFocusedElementIndex();
+
+    if (currentIndex === -1) {
+      currentIndex = this.adapter_.getListItemIndex(this.getListItem_(evt.target));
+
+      if (currentIndex < 0) {
+        // If this event doesn't have a mdc-list-item ancestor from the
+        // current list (not from a sublist), return early.
+        return;
+      }
+    }
+
+    if ((this.isVertical_ && arrowDown) || (!this.isVertical_ && arrowRight)) {
+      this.preventDefaultEvent_(evt);
+      this.focusNextElement(currentIndex);
+    } else if ((this.isVertical_ && arrowUp) || (!this.isVertical_ && arrowLeft)) {
+      this.preventDefaultEvent_(evt);
+      this.focusPrevElement(currentIndex);
+    } else if (isHome) {
+      this.preventDefaultEvent_(evt);
+      this.focusFirstElement();
+    } else if (isEnd) {
+      this.preventDefaultEvent_(evt);
+      this.focusLastElement();
+    }
+  }
+
+  /**
+   * Ensures that preventDefault is only called if the containing element doesn't
+   * consume the event, and it will cause an unintended scroll.
+   * @param {Event} evt
+   * @private
+   */
+  preventDefaultEvent_(evt) {
+    const tagName = ('' + evt.target.tagName).toLowerCase();
+    if (ELEMENTS_KEY_ALLOWED_IN.indexOf(tagName) === -1) {
+      evt.preventDefault();
+    }
+  }
+
+  /**
+   * Focuses the next element on the list.
+   * @param {Number} index
+   */
+  focusNextElement(index) {
+    const count = this.adapter_.getListItemCount();
+    let nextIndex = index + 1;
+    if (nextIndex >= count) {
+      if (this.wrapFocus_) {
+        nextIndex = 0;
+      } else {
+        // Return early because last item is already focused.
+        return;
+      }
+    }
+    this.adapter_.focusItemAtIndex(nextIndex);
+  }
+
+  /**
+   * Focuses the previous element on the list.
+   * @param {Number} index
+   */
+  focusPrevElement(index) {
+    let prevIndex = index - 1;
+    if (prevIndex < 0) {
+      if (this.wrapFocus_) {
+        prevIndex = this.adapter_.getListItemCount() - 1;
+      } else {
+        // Return early because first item is already focused.
+        return;
+      }
+    }
+    this.adapter_.focusItemAtIndex(prevIndex);
+  }
+
+  focusFirstElement() {
+    this.adapter_.focusItemAtIndex(0);
```

Comment:
do you need a check here to make sure there `this.adapter_.getListItemCount() > 0` ?